### PR TITLE
abstract instrumentation so that promiscuous blackhole can reuse it

### PR DESCRIPTION
**Pull request by benseligman**
This will make it really easy to extend newrelic instrumentation to promiscuous_black_hole. I added some basic tests, since the code is really abstract

### CODE REVIEW

File: lib/promiscuous/new_relic/instrumentation.rb [hunk 1]
```diff
@@ -1,5 +1,31 @@
 require 'new_relic/agent/instrumentation/controller_instrumentation'
 
+module PromiscuousNewRelicInstrumentedClass
+  extend ActiveSupport::Concern
+  include NewRelic::Agent::Instrumentation::ControllerInstrumentation
+
+  module ClassMethods
+    def instrument(method_name)
+      alias_method "orig_#{method_name}", method_name
```
Comment (kylessnell) [inline lib/promiscuous/new_relic/instrumentation.rb]: fancy

File: lib/promiscuous/new_relic/instrumentation.rb [hunk 2]
```diff
@@ -1,5 +1,31 @@
 require 'new_relic/agent/instrumentation/controller_instrumentation'
 
+module PromiscuousNewRelicInstrumentedClass
```
Comment (benseligman) [inline lib/promiscuous/new_relic/instrumentation.rb]: can probably be pulled out into its own file, so promiscuous_black_hole doesn't need to instrument the UnitOfWork class it's not using

File: lib/promiscuous/new_relic/instrumentation.rb [hunk 3]
```diff
@@ -1,5 +1,31 @@
...
+      end
+    end
+  end
+
+
+  included do
+    private
+    def namespace_for_rpm(operation)
+      "#{self.app}/#{operation.model.to_s}/#{operation.operation}"
+    end
```
Comment (kylessnell) [inline lib/promiscuous/new_relic/instrumentation.rb]: is this the new relic standard, or ours?

Comment (benseligman) [inline lib/promiscuous/new_relic/instrumentation.rb]: ours, afaik

Comment (kylessnell) [inline lib/promiscuous/new_relic/instrumentation.rb]: kinda sucks for an open source gem then.

**Comment by kylessnell** [comment]
LGTM

File: lib/promiscuous/new_relic/instrumentation.rb (hunk 3 above)
Comment (benseligman) [inline lib/promiscuous/new_relic/instrumentation.rb]: not sure what we can do about that. this is totally separate from promiscuous itself, so there's no obligation for a promiscuous user to use this

Comment (michaeldauria) [inline lib/promiscuous/new_relic/instrumentation.rb]: That comes from the promiscuous message, so it's ok

File: lib/promiscuous/new_relic/promiscuous_new_relic_instrumented.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+require 'new_relic/agent/instrumentation/controller_instrumentation'
+
+module PromiscuousNewRelicInstrumented
+  extend ActiveSupport::Concern
+  include NewRelic::Agent::Instrumentation::ControllerInstrumentation
+
+  module ClassMethods
+    def newrelic_namespace(&block)
+      if block_given?
+        @newrelic_namespace = block
```
Comment (kylessnell) [inline lib/promiscuous/new_relic/promiscuous_new_relic_instrumented.rb]: why a block and not a string?

File: spec/promiscuous/new_relic/promiscuous_new_relic_instrumented_spec.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    it 'calls perform_action_with_newrelic_trace with the right params' do
+      expect(subject).to receive(:perform_action_with_newrelic_trace).with(
+        :name => 'SuperCoolModel',
+        :class_name => 'Subscriber',
+        :force => true,
+        :category => 'OtherTransaction/Promiscuous'
+      )
+
+      subject.foo
+    end
```
Comment (kylessnell) [inline spec/promiscuous/new_relic/promiscuous_new_relic_instrumented_spec.rb:42]: nice

**Comment by kylessnell** [comment]
:+1: otherwise

File: lib/promiscuous/new_relic/instrumentation.rb [hunk 6]
```diff
@@ -8,24 +7,19 @@
...
-      def execute_operation(operation)
-        # We are not using the subscriber class name, because of polymorphism
-        # We only want the parent class basically
-        perform_action_with_newrelic_trace(:name => namespace_for_rpm(operation), :class_name => 'Subscriber', :force => true,
-                                           :category => "OtherTransaction/Promiscuous") do
-
-          execute_operation_without_rpm(operation)
-        end
+        instrument :process
+        newrelic_namespace { "Normcore2/#{message.base_type}/#{message.operation}" }
```
Comment (michaeldauria) [inline lib/promiscuous/new_relic/instrumentation.rb]: I don't like `Normcore2` here, that doesn't apply to others who are using the gem

**Pull request closed without merging**